### ISSUE TITLE
New settings editor cannot override bool value in workspace settings

### ISSUE THREAD

**Issue by yhvicey** [Microsoft/vscode#58038]
- VSCode Version: 1.27
- OS Version: Windows 10 1803

Steps to Reproduce:

1. Set some value to true in your user settings, like editor.formatOnSave:
![image](https://user-images.githubusercontent.com/8000820/45138329-94acaa80-b1de-11e8-8fa3-724d55bce045.png)

2. Open settings editor, switch to Workspace Settings, you will see the item is unchecked but showed (modified by user):
![image](https://user-images.githubusercontent.com/8000820/45138344-9d9d7c00-b1de-11e8-9de8-0f3c1b2ada4f.png)

3. If you want to set editor.formatOnSave to false in this workspace, there's no way to achieve this. If you check & uncheck the item, there will only be an empty settings.json file in .vscode folder:
![image](https://user-images.githubusercontent.com/8000820/45138389-cd4c8400-b1de-11e8-840a-08a1ad4a8167.png)

Does this issue occur when all extensions are disabled?: Yes

**Comment by roblourens** [comment]
That's correct - we decided to consider this an "advanced" feature that you can still configure in the json file, but not in the settings UI.

Showing a setting that's configured to its default value is a confusing UI experience, I'm still open to supporting this in the future though.

**Comment by yhvicey** [comment]
According to settings override order (Workspace Settings > User Setting > Default Settings), shouldn't workspace settings editor showing values inherited from user settings? If so, we can override values in workspace settings editor.

**Comment by roblourens** [comment]
I thought it would be less confusing to show always show the default unless a setting is modified in that scope.

**Comment by yhvicey** [comment]
Since there's a notice after the item (Modified in: User), I thought this won't be that confusing for users. These two settings (User and Workspace) are not parallel, so maybe we don't need to show default values in both editor?

**Comment by roblourens** [comment]
From #58901

> _(not sure if this is a feature request or a bug, since it was intentional)_
> 
> **Default settings may change.**  
> When users explicitly set a setting to a value, they expect to have a guarantee that the setting they chose explicitly will not change in the future, even if the default happens to change.
> 
> However, when you use the new settings editor to change a setting back to a value that happens to be the default, it removes the setting from the json.  
> This is not only wrong, but also unnecessary, since the settings editor already has a simple way to track modified settings and reset a setting to its default value:
> ![reset-setting](https://user-images.githubusercontent.com/38332912/45707170-881c4080-bb54-11e8-8c6e-089db0ac728e.png)
> 
> I strongly think the current behavior is incorrect, because these two things are completely different when we think about **user intention**:
> - explicitly wanting a value that happens to be the default
> - accepting whatever the default is for that setting
> 
> An example of why I think this sort of thing is important:  
> https://money.cnn.com/2018/06/07/technology/facebook-public-post-error/index.html
> It may be an extreme example, but you get the point.
> 
> This incorrect behavior is the sole reason I was still using the json to edit my settings, and it became more of a concern now that the new editor is the default.
> 
> The correct behavior should be:
> - When setting the value explicilty, mark the setting as modified/overriden and keep the setting in the json, even if it happens to be equal to the default. The user wants the guarantee that the setting will have THAT value, and a way to remember that such value was explicitly chosen by them.
> - When choosing "Reset Setting", remove the setting from the json, because the user wants to accept whatever the default is, even if it changes later. Mark the setting as not modified/overriden.
> 
> If possible, change the text and semantics of "modified" settings to "overriden" settings.

I agree that this is confusing. I think that supporting this for power users could also be confusing for newer users. Imagine if they click a checkbox on a bool setting to modify it, then click it again to "unmodify" it. Most people won't expect that action to have side effects, but it would have the side effect of writing the default value to the JSON.

I want to keep thinking about and discussing this, but I am also putting the experience for new users first in the UI.

**Comment by geekley** [comment]
Well, if a user checks and unchecks a setting, they will immediately notice that the setting now has a blue line, which wasn't there before. Even new users that don't know the exact meaning of the blue line will at least notice that something changed and didn't unchange.

If this wasn't intended, the first thing they will do (after possibly trying CTRL+Z) is look inside the gear icon. This is the intuitive thing to do, because its the only button related to the setting apart from the setting itself.

Even if they don't notice the change, or even if they don't get that the blue line means that there was some sort of change, I still believe strongly that the current behavior is not right. The side-effect of erasing a setting you had explicitly chosen before is more unexpected (even for experienced users) than the side-effect of adding and keeping a setting you explicitly just messed with.

Specially because there is **also** a separate reset button, so you wouldn't expect that explicitly setting something could also reset it.

**Comment by roblourens** [comment]
> Even new users that don't know the exact meaning of the blue line will at least notice that something changed and didn't unchange

Yes but that's not a good thing, because they will be confused.

> If this wasn't intended, the first thing they will do (after possibly trying CTRL+Z) is look inside the gear icon.

I disagree. I think many users will never notice or explore the context menu.

A checkbox is the simplest control in the world. It works the same on every form or in every settings page in every app, everywhere. If we change its behavior by introducing a mysterious third state, controlled by a different command in a hidden menu, we risk making users feel like they don't understand how a checkbox works. And if they don't understand even a checkbox, they will feel like they can't understand anything. This is why I want to proceed with caution.

**Comment by geekley** [comment]
Ok then, how about erasing the setting only if it wasn't already there before the settings tab was first open?
Then the current behaviour can be mantained, while people that already have a custom setting won't risk losing it unless they use the reset button.

**Comment by geekley** [comment]
My previous comment was for user settings.

In the specific case of this issue, (not being able to override the Workspace setting) [this solution](https://github.com/Microsoft/vscode/issues/50827#issuecomment-423004412) (an override checkbox before each setting) would solve it better, I think.

**Comment by roblourens** [comment]
@RedTailBullet had a good suggestion, that the default in the Workspace settings would match the configured value in the User settings. So here's how I interpret this:

- Starting clean with no configured settings
- Open User settings, see editor.insertSpaces is checked by default
- Open Workspace settings, editor.insertSpaces is checked by default
- Go back to User settings uncheck editor.insertSpaces. It gets the blue modified indicator
- Switch to workspace settings, find editor.insertSpaces, it appears unchecked by default, and doesn't have the modified indicator
- Click the checkbox, now it is checked and has the modified indicator

This is an interesting idea because it preserves "normal" checkbox behavior, it is transparent about the value at "this" scope (not taking the folder-level scope into account), and makes it easy and natural to override a value to the "default".

**Comment by RedTailBullet** [comment]
> @RedTailBullet had a good suggestion, that the default in the Workspace settings would match the configured value in the User settings. So here's how I interpret this:
> 
> * Starting clean with no configured settings
> * Open User settings, see editor.insertSpaces is checked by default
> * Open Workspace settings, editor.insertSpaces is checked by default
> * Go back to User settings uncheck editor.insertSpaces. It gets the blue modified indicator
> * Switch to workspace settings, find editor.insertSpaces, it appears unchecked by default, and doesn't have the modified indicator
> * Click the checkbox, now it is checked and has the modified indicator
> 
> This is an interesting idea because it preserves "normal" checkbox behavior, it is transparent about the value at "this" scope (not taking the folder-level scope into account), and makes it easy and natural to override a value to the "default".

Thanks roblourens! That's exactly what I mean.

**Comment by RedTailBullet** [comment]
One thing as well, the actual config.json for workspace setting should also be based on user settings.

For example:

* In user setting, the editor.insertSpaces is unchecked (which is not the default).
* In workspace setting, the editor.insertSpaces is checked again (which is the global default), and the json of the workspace setting actually have `editor.insertSpace: true`.

**Comment by SetTrend** [comment]
From the manual:

> Workspace settings override user settings.

See: [User and Workspace settings](https://code.visualstudio.com/docs/getstarted/settings)

**Comment by lucasrcosta** [comment]
I agree with @SetTrend: "user settings differ from user to user, whereas workspace settings don't".
I'm having an issue with "Trim Trailing Whitespace". I like to used it in all my projects, but if you want to contribute in a project with some whitespaces and you save, a lot of lines have changed. Right now my only option is to disable it on the User settings, hope I don't forget to turn it on again...

**Comment by ZSkycat** [comment]
Setting the workspace to the default value does not override the user settings. This is a very confusing behavior.
When I use the UI settings, if I don't pay attention, I will break my workspace settings

The UI settings of the workspace are displayed as inherited values will be better.

**Comment by EskelCz** [comment]
I just ran into this nonsense. :( Workspace settings must override user settings, at least for linters.

**Comment by esr360** [comment]
I'm also confused. My workspace settings should surely override my user settings?? Why wouldn't this be the expected behaviour?

**Comment by AlexandreDecan** [comment]
I got into this issue as well while testing the "remote - ssh" extension. On my laptop, I've a user setting that defines where to find linters in Python (e.g. path to flake8). On the remote machine, I would like to define this path to "flake8" for this workspace, but I can't because it's the default value assumed by vscode. I currently have no way to specify that for this specific workspace, I would like "flake8" (the default value in vscode) to override the value I put in my user settings.

**Comment by yhvicey** [comment]
> I got into this issue as well while testing the "remote - ssh" extension. On my laptop, I've a user setting that defines where to find linters in Python (e.g. path to flake8). On the remote machine, I would like to define this path to "flake8" for this workspace, but I can't because it's the default value assumed by vscode. I currently have no "direct way" (using the UI) to specify that for this specific workspace, I would like "flake8" (the default value in vscode) to override the value I put in my user settings.

To workaround you can change the setting editor to split json mode by setting below two options:

```json
  "workbench.settings.editor": "json",
  "workbench.settings.useSplitJSON": true,
```

Then you can override your settings in workspace tab manually:

![image](https://user-images.githubusercontent.com/8000820/64579251-c4276180-d3b4-11e9-9cb3-a8b9cba3e8bb.png)

**Comment by AlexandreDecan** [comment]
Thanks @yhvicey !

**Comment by roblourens** [comment]
You don't have to use the deprecated "split" json editor for this. Use the command "Open Settings (JSON)" or the button in the corner of the settings GUI that opens the backing JSON file.

**Comment by richard-biarri** [comment]
Can't override a setting to turn off format-on-save in Workspace settings if it's set to on in User settings. Blue "overrides User settings" line appears when you *check* the box (which is *wrong* because the override is the same as the User setting) but it disappears when you *uncheck* the box, which is now a different setting from the User setting.

If the Workspace is inheriting a User setting then that setting should be displayed in the Workspace settings unless it's overridden. A checked User setting should appear checked in Workspace, and when the unchecked, it should have the blue line indicating that it differs.

**Comment by sebastian-fredriksson-bernholtz** [comment]
@roblourens Don't you think it's more confusing to show a value that doesn't apply in Workspace Settings (UI) than to explicitly setting a value to the same as default? Considering that Workspace Settings are supposed to override all other settings?

I haven't looked at the code but it seems that:
1. Workspace Settings (UI) will show the default value unless specifically set in Workspace Settings (JSON). Regardless of whether the User Settings are overriding the Default Settings. 
2. Workspace Settings (UI) will also only write to Workspace Settings (JSON) if the value is different from the default value. It will delete the setting from Workspace Settings (JSON) if it is the same as Default.

The first issue is handled to some degree by `(Modified in: User)`. Considering Workspace Settings are supposed to override User Settings it's not very intuitive that being modified in User Settings means it overrides the setting shown in Workspace Settings (UI). It would be clearer if it showed `(Overridden by: User)`. Even better would be to show the actual value applied to the workspace and where it's been set. As in `(Set in: User)`/`(Set in: Default)`. For text input settings the applied value could be set as placeholder rather than value. 

I think the second issue is worse. Why make it make it impossible to make the workspace explicitly use the default setting through Workspace Settings (UI)? The only benefit with the current behaviour seems to be that it is possible to disable the Workspace Setting so that the User Setting is applied.

**Comment by k3n** [comment]
This violates the principle of least surprise IMHO.

**Comment by esr360** [comment]
How can I keep my personal projects at 2 space tabs whilst also allowing work projects to have 4 space tabs?

**Comment by andrewmiroshnichenko** [comment]
I just ran into this issue. From my point of view, on https://code.visualstudio.com/docs/getstarted/settings text
```
Workspace settings override user settings.
```
should be removed, because it is simply not true.

Can someone give approximate estimate for this issue resolution?

**Comment by heshuimu** [comment]
+1

**Comment by mattgaspar** [comment]
Currently to force the default value to override a user setting for a workspace you need to:
1. Click the gear icon next to the setting
2. Click "Copy Setting as JSON"
3. Open the **workspace's** settings.json file
4. Paste the default setting (that was copied in step 2) between the brackets

**Proposed Solution**
Add an additional option to the gear menu that says something like "Force default value" or have clickable text next to the setting (if the default is selected) that says "Force default" or "Override user setting".
Clicking any of these would be a shortcut to doing the manual steps above.

Also it would be helpful to show the actual active setting.  
How about if the active value for a setting does not equal the value you are looking at, have a message like _`Active value is currently: onFocusChange`_ appear next to or under the setting.

**Comment by 1valdis** [comment]
I agree, this is confusing. Please make workspace overriding more user-friendly. I was just trying to modify workspace settings and wondering why this doesn't work at all and googled to find this issue.

**Comment by Drixled** [comment]
Is there any ETA for this? Even when I paste the JSON on my settings.json like @mattgaspar said the change does not override because the hierarchy rule is Workspace > User.

**Comment by shamilovtim** [comment]
This has been driving me nuts and I've probably sunk at least 20-30 minutes over the course of several months figuring out what VSCode was doing. I had **no idea** the setting wasn't actually being pushed to the JSON. It's not best practice for a UI to randomly make an opaque exception like this, not do what the user is asking it to do, and provide no feedback about why it did that. Perhaps you guys could show an error message saying "Setting change was not successful. This setting needs to be overriden directly in the JSON because of _____." I also find it strange that this particular functionality is hostile toward UI settings users but not JSON settings users. Why? That's weird.

**Comment by SetTrend** [comment]
We have more than 20 upvotes on this issue now. Just like with [this issue](104183).

What is the further process of this improvement request? Is the "20 upvotes limit" just a mirage?

**Comment by gjsjohnmurray** [comment]
> We have more than 20 upvotes on this issue now. Just like with [this issue](104183).
> 
> What is the further process of this improvement request? Is the "20 upvotes limit" just a mirage?

@SetTrend the 20 upvote thing is only relevant for an issue that was classified by a team member as 'Backlog Candidate'. Then, getting 20 upvotes on the initial post within 60 days would move it automatically to 'Backlog'. Still no guarantee of implementation, but arguably better than being in limbo, tagged as 'feature-request'.

@roblourens please move this to Backlog. Judging by the number of other issues that have referenced it over the years, this behaviour continues to confuse and surprise people.

**Comment by kasperskov909** [comment]
<<< Look at this clown thinking his workspace settings were enforced on his team members. 

Please fix this :)

**Comment by ShadowLNC** [comment]
Coupled with [vscode-python#7314](https://github.com/microsoft/vscode-python/issues/7314), it is _impossible_ to enable Pylint without manually editing settings.json. Allowing user-toggled settings to persist would mean I would have never encountered the other issue (as I toggled the Pylint setting prior). It also protects against similar bugs where a default is not correctly interpreted, providing the user toggles the setting.

Replying to https://github.com/microsoft/vscode/issues/58038#issuecomment-423002429

> I disagree. I think many users will never notice or explore the context menu.

I know I am not "many users", but I personally did look at the setting to reset values - without realising that VSCode doesn't allow you to explicitly set the default value without editing settings.json. As suggested by others, a "reset" button (e.g. the revert circular arrow) coupled with the already-present blue bar would be more than sufficient in my opinion.

Also _extremely_ confusing to me was the fact that Workspace settings, while inherited from User settings, display the VSCode default if inherited from User settings, rather than the User value if defined. The `Modified in: User` label suggests it's displaying the value of the User setting, but this is not the case.

**Comment by camlafit** [comment]
Hello

I confirm this behavior is not logical, as configuration are cascaded then default value depend to hierarchic value not to main default value.
In my case I spent some time to use some option set in my project to my workspace. I've not understood how enable correct option and edit json file is not a friendly feature when anywhere we can edit it. Option displayed was not option used by vscode in the context. We get then an incoherence.

Could be nice to display config value following cascading context and be coherent.

Thanks a lot :)

**Comment by beauxq** [comment]
> I thought it would be less confusing to show always show the default unless a setting is modified in that scope.

This is where this all went wrong. I don't see any reasoning to back up this idea.

If a user cares what the default is, they will look for that information somewhere else (some default label next to something - not in the same place the setting is set).

Most of the time, if a user is in the settings editor, they don't even care what the default is. They just know what they want and are trying to set it to that.

There is no reason to be showing the default in the place where the setting is set.

If it isn't modified in this scope, it should show whatever is inherited.

If you care about showing the default, put that somewhere else.

**Comment by glasser** [comment]
Whether or not the behavior is changed here, I think the docs could be improved. In https://code.visualstudio.com/docs/getstarted/settings it says:

> Workspace settings override user settings. Workspace settings are specific to a project and can be shared across developers on a project.

It seems like it would helpful to add a sentence or two, perhaps as a note, saying something like:

> Note: you cannot use the settings editor create a workspace setting that matches the setting's default value. If you are trying to ensure that when in a workspace, a particular setting has its default value even if users configure it, you must add it to the workspace settings file manually.

**Comment by DevilXD** [comment]
This one is +1 from me. I have the Python extension configured to lint files with MyPy at the User scope (with custom arguments list), and I thus expected these settings to apply to every new project I'd open, and I was honestly surprised to learn that "Workspace scope overwriting User scope" would pretty much revert these settings back to default, forcing me to repeat the exact same configuration in the current workspace, and any other workspace I'd open. I don't think that's how it should work.

I'm fine with workspace settings overwriting user settings, but if there's no workspace setting configured, user scope settings should be applied instead of the default.

**Comment by jaklan** [comment]
@DevilXD it works exactly in a way you described, you don't need to repeat any configuration. This issue is about **not** being able to revert default settings in the Workspace scope, when modified in the User scope, from the Settings UI.

**Comment by DevilXD** [comment]
@jaklan For some reason it didn't do so, until I set them in the workspace. Not sure what that was about then.

**Comment by glasser** [comment]
> Whether or not the behavior is changed here, I think the docs could be improved.

I've opened a docs PR with the above suggestion: https://github.com/microsoft/vscode-docs/pull/5190

**Comment by shamilovtim** [comment]
When will this be fixed?

**Comment by rzhao271** [comment]
Starting from tomorrow's [Insiders](https://code.visualstudio.com/insiders/) release, users should be able to add settings to the workspace file that stay there until manually reset.

The default values the settings display in the event that they have also been modified in the user settings will still be the default values rather than the user values for now, though.

**Comment by mind6** [comment]
If I set the User setting tab size to 3 (a non-default value), the Workspace setting shows 4 (the default value) in the UI, but behaves as if tab size is 3 in the editor.

My desire is to use 4 the default value for Workspace, but keep 3 in the User setting. The workaround is to manually put
```
{...
	"settings": {
		"editor.detectIndentation": false,
		"editor.tabSize": 4
...}
```
In the workspace settings files.

Is this caused by the same issue?

**Comment by jepperaskdk** [comment]
What's the status on this?
I would expect workspace settings to override user settings (how else can we enforce any settings for the team working on a codebase?). If you're worried about this being a breaking change, how about releasing it as an opt-in setting initially?

**Comment by ShadowLNC** [comment]
At the risk of being a "+1", I ran into this again due to https://github.com/microsoft/vscode-mypy/issues/184 (I originally hit this in June 2021 with Pylint).

Workspace settings UI continued to show me it was set to `useBundled` even though my User settings are `fromEnvironment` (the latter causing the extension to error out due to the linked issue).

At least now I can change the setting to `fromEnvironment` and then back to `useBundled` now and it will persist, rather than having to edit the JSON - thanks! 😅

**Comment by happyme531** [comment]
This issue still exists for remote scope.

**Comment by ParadoxV5** [comment]
This issue also impacts extension-contributed default settings.
